src/server.rs

```diff
@@ -1,105 +1,285 @@
-            let mut re = ResponseException::new();
-            re.set_error(String::from("Unrecognized request"));
-            response.set_exception(re)
+        Request_oneof_value::flush(_) => {
+            consensus.flush().await;
+            Response_oneof_value::flush(ResponseFlush::new())
         }
-    }
+        Request_oneof_value::info(request) => {
+            Response_oneof_value::info(info.info(request.into()).await.into())
+        }
+        Request_oneof_value::set_option(request) => {
+            Response_oneof_value::set_option(info.set_option(request.into()).await.into())
+        }
+        Request_oneof_value::init_chain(request) => {
+            consensus_state
+                .lock()
+                .await
+                .validate(ConsensusState::InitChain);
+            Response_oneof_value::init_chain(consensus.init_chain(request.into()).await.into())
+        }
+        Request_oneof_value::query(request) => {
+            Response_oneof_value::query(info.query(request.into()).await.into())
+        }
+        Request_oneof_value::begin_block(request) => {
+            consensus_state
+                .lock()
+                .await
+                .validate(ConsensusState::BeginBlock);
+            Response_oneof_value::begin_block(consensus.begin_block(request.into()).await.into())
+        }
+        Request_oneof_value::check_tx(request) => {
+            Response_oneof_value::check_tx(mempool.check_tx(request.into()).await.into())
+        }
+        Request_oneof_value::deliver_tx(request) => {
+            consensus_state
+                .lock()
+                .await
+                .validate(ConsensusState::DeliverTx);
+            Response_oneof_value::deliver_tx(consensus.deliver_tx(request.into()).await.into())
+        }
+        Request_oneof_value::end_block(request) => {
+            consensus_state
+                .lock()
+                .await
+                .validate(ConsensusState::EndBlock);
+            Response_oneof_value::end_block(consensus.end_block(request.into()).await.into())
+        }
+        Request_oneof_value::commit(_) => {
+            consensus_state
+                .lock()
+                .await
+                .validate(ConsensusState::Commit);
+            Response_oneof_value::commit(consensus.commit().await.into())
+        }
+    };
+
+    let mut response = Response::new();
+    response.value = Some(value);
+
+    log::debug!("Sending response: {:?}", response);
+
     response
 }
+
+#[derive(Debug, Clone, Copy)]
+pub enum ConsensusState {
+    InitChain,
+    BeginBlock,
+    DeliverTx,
+    EndBlock,
+    Commit,
+}
+
+impl Default for ConsensusState {
+    #[inline]
+    fn default() -> Self {
+        ConsensusState::InitChain
+    }
+}
+
+impl ConsensusState {
+    pub fn validate(&mut self, mut next: ConsensusState) {
+        let is_valid = match (&self, next) {
+            (ConsensusState::InitChain, ConsensusState::InitChain) => true,
```

Comment:
as discussed, this is not a valid transition.

There are also more invariants around consensus state -- e.g. that block height of a new block is previous_block_height+1.

Perhaps this validation could be left out for now (it's not mentioned in the changelog?)

Reply:
Changed. Now `consensus_state` is stored in a file when `InitChain` is called and in case of a server restart, it is read from a previously stored file.

Reply:
That's a bit inflexible -- perhaps just remove the validation for now?

It may be preferable to have a stateless validation (e.g. that will kick-in after initchain or begin block with certain height), or have `Server::run` take it?

Reply:
Why do you want to remove the validation? Maybe we can revert back to the one we already had (it looks like a hack but it works)?

Reply:
Removed all the persistence changes and added `is_initialized: bool` in `Server::new`.